File: java/src/test/java/br/com/vps/consulting/b2b/management/shared/core/vo/MoneyTest.java
package br.com.vps.consulting.b2b.management.shared.core.vo;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MoneyTest {

    @Test
    @DisplayName("Should create Money with explicit value and currency")
    void shouldCreateWithAmountAndCurrency() {
        final var money = new Money(new BigDecimal("100.00"), "USD");
        assertThat(money.value()).isEqualByComparingTo("100.00");
        assertThat(money.currency()).isEqualTo("USD");
    }

    @Test
    @DisplayName("Should create Money with default BRL currency via of(BigDecimal)")
    void shouldCreateWithDefaultCurrencyFromBigDecimal() {
        final var money = Money.of(new BigDecimal("50.00"));
        assertThat(money.currency()).isEqualTo(Money.DEFAULT_CURRENCY);
        assertThat(money.value()).isEqualByComparingTo("50.00");
    }

    @Test
    @DisplayName("Should create Money with default BRL currency via of(String)")
    void shouldCreateWithDefaultCurrencyFromString() {
        final var money = Money.of("75.50");
        assertThat(money.currency()).isEqualTo(Money.DEFAULT_CURRENCY);
        assertThat(money.value()).isEqualByComparingTo("75.50");
    }

    @Test
    @DisplayName("Should allow zero value")
    void shouldAllowZeroAmount() {
        final var money = new Money(BigDecimal.ZERO, "BRL");
        assertThat(money.isZero()).isTrue();
    }

    @Test
    @DisplayName("Should reject null value")
    void shouldRejectNullAmount() {
        assertThrows(NullPointerException.class, () -> new Money(null, "BRL"));
    }

    @Test
    @DisplayName("Should reject null currency")
    void shouldRejectNullCurrency() {
        assertThrows(NullPointerException.class, () -> new Money(new BigDecimal("10.00"), null));
    }

    @Test
    @DisplayName("Should reject negative value")
    void shouldRejectNegativeAmount() {
        assertThatThrownBy(() -> new Money(new BigDecimal("-0.01"), "BRL"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("cannot be negative");
    }

    @Test
    @DisplayName("Should add two Money values with same currency")
    void shouldAddTwoMoneyWithSameCurrency() {
        final var a = new Money(new BigDecimal("30.00"), "BRL");
        final var b = new Money(new BigDecimal("20.00"), "BRL");
        final var result = a.add(b);
        assertThat(result.value()).isEqualByComparingTo("50.00");
        assertThat(result.currency()).isEqualTo("BRL");
    }

    @Test
    @DisplayName("Should reject addition of Money with different currencies")
    void shouldRejectAdditionWithDifferentCurrencies() {
        final var brl = new Money(new BigDecimal("10.00"), "BRL");
        final var usd = new Money(new BigDecimal("10.00"), "USD");
        assertThatThrownBy(() -> brl.add(usd))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Currency mismatch");
    }

    @Test
    @DisplayName("Should multiply Money by positive quantity")
    void shouldMultiplyByPositiveQuantity() {
        final var price = new Money(new BigDecimal("15.00"), "BRL");
        final var result = price.multiply(3);
        assertThat(result.value()).isEqualByComparingTo("45.00");
        assertThat(result.currency()).isEqualTo("BRL");
    }

    @Test
    @DisplayName("Should return zero when multiplying by zero")
    void shouldReturnZeroWhenMultiplyingByZero() {
        final var price = new Money(new BigDecimal("20.00"), "BRL");
        final var result = price.multiply(0);
        assertThat(result.isZero()).isTrue();
    }

    @Test
    @DisplayName("Should reject negative quantity in multiply")
    void shouldRejectNegativeQuantityInMultiply() {
        final var price = new Money(new BigDecimal("10.00"), "BRL");
        assertThatThrownBy(() -> price.multiply(-1))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Quantity cannot be negative");
    }

    @Test
    @DisplayName("Should return true when value is greater than other")
    void shouldReturnTrueWhenIsGreaterThan() {
        final var larger = new Money(new BigDecimal("100.00"), "BRL");
        final var smaller = new Money(new BigDecimal("50.00"), "BRL");
        assertThat(larger.isGreaterThan(smaller)).isTrue();
    }

    @Test
    @DisplayName("Should return false when value is not greater than other")
    void shouldReturnFalseWhenNotGreaterThan() {
        final var a = new Money(new BigDecimal("50.00"), "BRL");
        final var b = new Money(new BigDecimal("50.00"), "BRL");
        assertThat(a.isGreaterThan(b)).isFalse();
    }

    @Test
    @DisplayName("Should reject isGreaterThan comparison with different currency")
    void shouldRejectIsGreaterThanWithDifferentCurrency() {
        var brl = new Money(new BigDecimal("100.00"), "BRL");
        var usd = new Money(new BigDecimal("50.00"), "USD");
        assertThatThrownBy(() -> brl.isGreaterThan(usd))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Currency mismatch");
    }

    @Test
    @DisplayName("Should return true for isZero when value is zero")
    void shouldReturnTrueForIsZero() {
        var money = new Money(BigDecimal.ZERO, "BRL");
        assertThat(money.isZero()).isTrue();
    }

    @Test
    @DisplayName("Should return false for isZero when value is positive")
    void shouldReturnFalseForIsZeroWhenPositive() {
        var money = new Money(new BigDecimal("0.01"), "BRL");
        assertThat(money.isZero()).isFalse();
    }
}

File: kotlin/src/test/kotlin/br/com/vps/consulting/b2b/management/shared/core/vo/MoneyTest.kt
package br.com.vps.consulting.b2b.management.shared.core.vo

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import java.math.BigDecimal

class MoneyTest {

    @Test
    fun `should create Money with default BRL currency via of(BigDecimal)`() {
        val money = Money.of(BigDecimal("100.00"))

        assertThat(money.value).isEqualByComparingTo("100.00")
        assertThat(money.currency).isEqualTo("BRL")
    }

    @Test
    fun `should create Money with default BRL currency via of(Long)`() {
        val money = Money.of(100L)

        assertThat(money.value).isEqualByComparingTo("100")
        assertThat(money.currency).isEqualTo("BRL")
    }

    @Test
    fun `should add two Money values with same currency`() {
        val result = Money.of(BigDecimal("10.00")) + Money.of(BigDecimal("5.00"))

        assertThat(result.value).isEqualByComparingTo("15.00")
    }

    @Test
    fun `should reject addition of Money with different currencies`() {
        val brl = Money(BigDecimal("10.00"), "BRL")
        val usd = Money(BigDecimal("5.00"), "USD")

        assertThatThrownBy { brl + usd }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("moedas diferentes")
    }

    @Test
    fun `should subtract two Money values with same currency`() {
        val result = Money.of(BigDecimal("10.00")) - Money.of(BigDecimal("4.00"))

        assertThat(result.value).isEqualByComparingTo("6.00")
    }

    @Test
    fun `should reject subtraction with different currencies`() {
        val brl = Money(BigDecimal("10.00"), "BRL")
        val usd = Money(BigDecimal("5.00"), "USD")

        assertThatThrownBy { brl - usd }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("moedas diferentes")
    }

    @Test
    fun `should multiply Money by a positive factor`() {
        val result = Money.of(BigDecimal("10.00")) * 3

        assertThat(result.value).isEqualByComparingTo("30.00")
    }

    @Test
    fun `should return zero-value Money when multiplying by zero`() {
        val result = Money.of(BigDecimal("10.00")) * 0

        assertThat(result.value).isEqualByComparingTo("0")
    }

    @Test
    fun `should return true when isGreaterThan and value is strictly greater with same currency`() {
        val greater = Money.of(BigDecimal("10.00"))
        val smaller = Money.of(BigDecimal("5.00"))

        assertThat(greater.isGreaterThan(smaller)).isTrue()
    }

    @Test
    fun `should return false when isGreaterThan and values are equal`() {
        val first = Money.of(BigDecimal("10.00"))
        val second = Money.of(BigDecimal("10.00"))

        assertThat(first.isGreaterThan(second)).isFalse()
    }

    @Test
    fun `should reject isGreaterThan comparison with different currency`() {
        val brl = Money(BigDecimal("10.00"), "BRL")
        val usd = Money(BigDecimal("5.00"), "USD")

        assertThatThrownBy { brl.isGreaterThan(usd) }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("moedas diferentes")
    }

    @Test
    fun `should return true for isNegativeOrZero when value is zero`() {
        assertThat(Money.of(BigDecimal.ZERO).isNegativeOrZero()).isTrue()
    }

    @Test
    fun `should return true for isNegativeOrZero when value is negative`() {
        assertThat(Money.of(BigDecimal("-1.00")).isNegativeOrZero()).isTrue()
    }

    @Test
    fun `should return false for isNegativeOrZero when value is positive`() {
        assertThat(Money.of(BigDecimal("0.01")).isNegativeOrZero()).isFalse()
    }

}
